src/util/buf.rs

```diff
@@ -11,93 +11,429 @@
+            self.end - self.start
+        } else {
+            self.buf.cap() - self.start + self.end
+        }
+    }
 
-            if count == left.len() {
-                if let Some(n) = try!(w.try_write(right)) {
-                    count += n;
-                }
-            }
+    #[inline]
+    unsafe fn buf_as_slice(&self) -> &[u8] {
+        slice::from_raw_parts(self.buf.ptr(), self.buf.cap())
+    }
 
-            count
-        };
+    #[inline]
+    unsafe fn buf_as_slice_mut(&self) -> &mut [u8] {
+        slice::from_raw_parts_mut(self.buf.ptr(), self.buf.cap())
+    }
 
-        self.buf.drain(0..count);
-        Ok(count)
+    #[inline]
+    pub fn capacity(&self) -> usize {
+        self.buf.cap() - 1
     }
 
-    pub fn len(&self) -> usize {
-        self.buf.len()
+    #[inline]
+    pub fn is_full(&self) -> bool {
+        self.len() == self.capacity()
     }
 
+    #[inline]
     pub fn is_empty(&self) -> bool {
         self.len() == 0
     }
+
+    /// Get the written buf.
+    fn slice(&self) -> (&[u8], &[u8]) {
+        unsafe {
+            let buf = self.buf_as_slice();
+            if self.end >= self.start {
+                (&buf[self.start..self.end], &[])
+            } else {
+                (&buf[self.start..], &buf[..self.end])
+            }
+        }
+    }
+
+    /// Get the not written buf.
+    fn slice_append(&mut self) -> (&mut [u8], &mut [u8]) {
+        if self.is_full() {
+            return (&mut [], &mut []);
+        }
+        unsafe {
+            let start = self.start;
+            let end = self.end;
+            let cap = self.capacity();
+            let buf = self.buf_as_slice_mut();
+
+            if start == 0 {
+                (&mut buf[end..cap], &mut [])
+            } else if start <= end {
+                let (right, left) = buf.split_at_mut(end);
+                let (right, _) = right.split_at_mut(start - 1);
+                (left, right)
+            } else {
+                (&mut buf[end..start - 1], &mut [])
+            }
+        }
+    }
+
+    /// Ensure the capacity of inner buf not less than `capacity`.
+    ///
+    /// If capacity is larger than inner buf, a larger buffer will be reallocated.
+    /// Allocated buffer's capacity doesn't have to be equal to specified value.
+    pub fn ensure(&mut self, capacity: usize) {
+        if capacity <= self.capacity() {
+            return;
+        }
+
+        let cap = self.buf.cap();
+        self.buf.reserve(cap, capacity + 1 - cap);
+        let new_cap = self.buf.cap();
+
+        // After the buf being extended, we have to make the written data layout correctly.
+        unsafe {
+            if self.start <= self.end {
+                // written data are linear, no need to move.
+                return;
+            } else if new_cap - cap > self.end && self.end <= cap - self.start {
+                // left part can be fit in new buf and is shorter.
+                // [ll...rrr...] -> [.....rrrll.]
```

Comment:
You mean `[.....rrrll.]` better than `[ll....rrr]`? Because it move less data. Former only moves 2 bytes, latter moves 3 bytes.